pkg/controllers/user/multiclusterapp/controller.go

```diff
@@ -0,0 +1,267 @@
+				Name:        appNameInProject,
+				Namespace:   projectNS,
+				Annotations: ann,
+				Labels:      set,
+			},
+			Spec: pv3.AppSpec{
+				ProjectName:         t.ProjectName,
+				TargetNamespace:     targetNamespace,
+				ExternalID:          externalID,
+				MultiClusterAppName: mcapp.Name,
+			},
+		}
+
+		if ans, ok := answerMap[t.ProjectName]; ok {
+			app.Spec.Answers = ans
+			answerFound = true
+		}
+
+		if !answerFound {
+			// find the answers for the cluster of this project
+			split := strings.SplitN(t.ProjectName, ":", 2)
+			if len(split) != 2 {
+				return mcapp, fmt.Errorf("error in splitting project name")
+			}
+			clusterName := split[0]
+			if ans, ok := answerMap[clusterName]; ok {
+				app.Spec.Answers = ans
+				answerFound = true
+			}
+
+			if !answerFound {
+				if ans, ok := answerMap[globalScopeAnswersKey]; ok {
+					app.Spec.Answers = ans
+				}
+			}
+		}
+
+		// Create the target namespace first
+		// Check if the target namespace already exists, if not then continue, if it does then add a random 5-character string at the end to make it unique
+		// Adding the projectId as an annotation is necessary, else the API/UI and UI won't list any of the resources from this namespace
+		nsLabel := make(map[string]string)
+		nsLabel[projectIDFieldLabel] = projectNS
+		nsAnnotation := make(map[string]string)
+		nsAnnotation[projectIDFieldLabel] = t.ProjectName
+		nsAnnotation[creatorIDAnn] = creatorID
+		n := v1.Namespace{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:        targetNamespace,
+				Labels:      nsLabel,
+				Annotations: nsAnnotation,
+			},
+		}
+		_, err = m.NsClient.Get(targetNamespace, metav1.GetOptions{})
+		if err != nil {
+			if !apierrors.IsNotFound(err) {
+				return nil, err
+			}
+			// namespace doesn't already exist
+			_, err := m.NsClient.Create(&n)
+			if err != nil && !apierrors.IsAlreadyExists(err) {
+				return mcapp, err
+			}
+		}
+
+		// Now create the App instance
+		_, err = m.AppClient.Create(&app)
+		if err != nil && !apierrors.IsAlreadyExists(err) {
+			return mcapp, err
+		}
+	}
+	return mcapp, nil
+}
+
+func (m *MCAppController) deleteApps(mcAppName string) (runtime.Object, error) {
+	// get all apps with label "multiClusterAppId" = name of this app
+	set := labels.Set(map[string]string{multiClusterAppIDSelector: mcAppName})
+	// delete all namespaces with the name
+
+	// to get all apps, get all clusters first, then get all apps in all projects of all clusters
+	appsToDelete := []*pv3.App{}
+	clusters, err := m.ClusterLister.List("", labels.NewSelector())
+	if err != nil {
+		return nil, err
+	}
+	for _, c := range clusters {
+		projects, err := m.ProjectLister.List(c.Name, labels.NewSelector())
+		if err != nil {
+			return nil, err
+		}
+		for _, p := range projects {
+			apps, err := m.AppLister.List(p.Name, set.AsSelector())
+			if err != nil {
+				return nil, err
+			}
+			appsToDelete = append(appsToDelete, apps...)
+		}
+	}
+
+	go func([]*pv3.App) {
```

Comment:
how do we handle the failures in the go routine? It is fine to schedule the removal in parallel, but you should fetch the result from all go routines, and return error to the framework if one of them failed.